WordPress/WordPressApi/WordPressComOAuthClient.m

```diff
@@ -66,67 +96,78 @@ - (void)requestOneTimeCodeWithUsername:(NSString *)username
+                      success:^(NSURLSessionDataTask *task, id responseObject) {
+                          if (success) {
+                              success();
+                          }
+                      } failure:^(NSURLSessionDataTask *task, NSError *error) {
+                          if ([task.response isKindOfClass:[NSHTTPURLResponse class]]) {
+                              NSHTTPURLResponse *httpURLResponse = (NSHTTPURLResponse *)task.response;
+                              error = [self processError:error forResponse:httpURLResponse];
+                          }
+
+                          // SORRY:
+                          // SMS Requests will still return WordPressComOAuthErrorNeedsMultifactorCode. In which case,
+                          // we should hit the success callback.
+                          if (error.code == WordPressComOAuthErrorNeedsMultifactorCode) {
+                              if (success) {
+                                  success();
+                              }
+                          } else if (failure) {
+                              failure(error);
+                          }
+                      }];
 }
 
-- (NSError *)processError:(NSError *)error forOperation:(AFHTTPRequestOperation *)operation {
-    if (operation.response.statusCode >= 400 && operation.response.statusCode < 500) {
-        // Bad request, look for errors in the JSON response
-		NSDictionary* response = operation.responseObject;
-
-        if (response) {
-            NSString *errorCode = [response stringForKey:@"error"];
-            NSString *errorDescription = [response stringForKey:@"error_description"];
-
-            /*
-             Possible errors:
-             - invalid_client: client_id is missing or wrong, it shouldn't happen
-             - unsupported_grant_type: client_id doesn't support password grants
-             - invalid_request: A required field is missing/malformed
-             - invalid_request: Authentication failed
-             - needs_2fa: Multifactor Authentication code is required
-             */
-            
-            NSDictionary *errorsMap = @{
-                @"invalid_client"           : @(WordPressComOAuthErrorInvalidClient),
-                @"unsupported_grant_type"   : @(WordPressComOAuthErrorUnsupportedGrantType),
-                @"invalid_request"          : @(WordPressComOAuthErrorInvalidRequest),
-                @"needs_2fa"                : @(WordPressComOAuthErrorNeedsMultifactorCode)
-            };
-
-            NSNumber *mappedCode = errorsMap[errorCode] ?: @(WordPressComOAuthErrorUnknown);
-            
-            return [NSError errorWithDomain:WordPressComOAuthErrorDomain
-                                       code:mappedCode.intValue
-                                   userInfo:@{NSLocalizedDescriptionKey: errorDescription}];
-        }
+- (NSError *)processError:(NSError *)error forResponse:(NSHTTPURLResponse *)response {
```

Comment:
Probably out of scope for this, but I'm wondering if all this error handling should be inside a custom response serializer instead.